--- a/Source/WTF/wtf/Platform.h	2014-03-02 11:35:50.845189616 +0200
+++ b/Source/WTF/wtf/Platform.h	2014-03-02 11:47:47.173698127 +0200
@@ -149,6 +149,11 @@
 #if   defined(__x86_64__) \
     || defined(_M_X64)
 #define WTF_CPU_X86_64 1
+
+#if defined(__ILP32__)
+#define WTF_CPU_X86_X32 1
+#endif
+
 #endif
 
 /* CPU(ARM64) - Apple */
@@ -636,7 +641,7 @@
 #endif
 
 #if !defined(WTF_USE_JSVALUE64) && !defined(WTF_USE_JSVALUE32_64)
-#if (CPU(X86_64) && (OS(UNIX) || OS(WINDOWS))) \
+#if (CPU(X86_64) && (OS(UNIX) || OS(WINDOWS)) && !CPU(X86_X32)) \
     || (CPU(IA64) && !CPU(IA64_32)) \
     || CPU(ALPHA) \
     || CPU(ARM64) \
@@ -655,6 +660,12 @@
 #define ENABLE_FTL_JIT 0
 #endif
 
+#if !defined(ENABLE_JIT) && CPU(X86_X32)
+#define ENABLE_JIT 0
+#define ENABLE_YARR_JIT 0
+#define ENABLE_ASSEMBLER 0
+#endif
+
 /* Disable the JIT on versions of GCC prior to 4.1 */
 #if !defined(ENABLE_JIT) && COMPILER(GCC) && !GCC_VERSION_AT_LEAST(4, 1, 0)
 #define ENABLE_JIT 0
--- a/Source/WTF/wtf/Atomics.h	2014-03-02 14:13:53.223343273 +0200
+++ b/Source/WTF/wtf/Atomics.h	2014-03-02 14:14:34.613636071 +0200
@@ -146,7 +146,11 @@
 #if CPU(X86_64)
     bool result;
     asm volatile(
+#ifdef __ILP32__
+        "lock; cmpxchg %3, %2\n\t"
+#else
         "lock; cmpxchgq %3, %2\n\t"
+#endif
         "sete %1"
         : "+a"(expected), "=q"(result), "+m"(*location)
         : "r"(newValue)
